File: #0074.search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {

        int rowSize = matrix.size();
        int colSize = matrix[0].size();
        int index = 0;
        int boardSize = rowSize * colSize - 1;
        int middle;
        while (index <= boardSize){
            middle = (index + boardSize) / 2;
            if (matrix[middle / colSize][middle % colSize] == target) {
                return true;
            } else if(matrix[middle / colSize][middle % colSize] < target) {
                index = middle + 1;
            } else{
                boardSize = middle - 1;
            }
        }
        return false;
    }
};

File: #0074.search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    let low = 0;
    let high = matrix.length - 1;
    let mid = 0;
    while (low <= high) {
        mid = Math.ceil((high + low) / 2);

        if (matrix[mid].includes(target)) {
            return true;
        }

        let lastMid = matrix[mid][matrix[mid].length - 1]
        if (lastMid < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return matrix[mid].includes(target);
};
